bindings/rust/s2n-tls/src/config.rs

```diff
@@ -749,8 +763,80 @@ impl Default for Context {
             private_key_callback: None,
             verify_host_callback: None,
             session_ticket_callback: None,
+            connection_initializer: None,
             wall_clock: None,
             monotonic_clock: None,
         }
     }
 }
+
+/// A trait executed before a new connection negotiates TLS.
+///
+/// Used for any dynamic configuration of the connection.
+/// Use in conjunction with
+/// [config::Builder::set_connection_initializer](`crate::config::Builder::set_connection_initializer()`).
+pub trait ConnectionInitializer: 'static + Send + Sync {
+    /// The application can return an `Ok(None)` to resolve the callback
+    /// synchronously or return an `Ok(Some(ConnectionFuture))` if it wants to
+    /// run some asynchronous task before resolving the callback.
+    ///
+    fn initialize_connection(
+        &self,
+        connection: &mut crate::connection::Connection,
+    ) -> ConnectionFutureResult;
+}
+
+impl<A: ConnectionInitializer, B: ConnectionInitializer> ConnectionInitializer for (A, B) {
+    fn initialize_connection(
+        &self,
+        connection: &mut crate::connection::Connection,
+    ) -> ConnectionFutureResult {
```

Comment:
The way our tokio bindings solve this problem (the TLSAcceptor can only take one "s2n-tls thing" to create all its different connections) is with [connection::Builder](https://github.com/aws/s2n-tls/blob/main/bindings/rust/s2n-tls/src/connection/builder.rs). It's a trait that connections, configs, connection pools, etc all implement that basically represents a thing that can produce at least one connection.

Maybe we should consider an Async version of the connection::Builder rather than a different initialization model altogether.

Reply:
Specifically, a [ModifiedBuilder](https://github.com/aws/s2n-tls/blob/main/bindings/rust/s2n-tls/src/connection/builder.rs#L40-L51) with a Config is basically what you want.